ITISFoundation/osparc-simcore
# ✨ Limit inflight requests to the api-server

## What do these changes do?
- Introduce a new traefik middleware which limits the total number of in-flight requests to the api-server. This means there is an absolute limit to the number of requests the api-server will be handling at any given time (see https://doc.traefik.io/traefik/middlewares/http/inflightreq/). 
- If the number of allowed in-flight requests to the api-server is exceeded traefik returns a [429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)
- Related ops MR: https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/651

## Related issue/s


- #5952 

## How to test



run local deployment with `API_SERVER_INFLIGHTREQ_AMOUNT=1` in `.env`. Then run
```bash
seq 10 | parallel -j 10 "curl -X 'GET'   'http://<ip address>.nip.io:<api-server-port>/v0/solvers'   -H 'accept: application/json'   -H 'Authorization: Basic <auth>' -v > output{#}.txt"
```
with `<api-server-port>`=the port exposing the api-server **through traefik** and `<auth>`=valid authorization.
Observe the `output*.txt` files and see that only a single request received a 2xx return code and the others received a 429.

## Dev-ops checklist
